compiler/arm/codegen/ARMSystemLinkage.cpp

```diff
@@ -167,156 +172,270 @@ TR::ARMLinkageProperties TR::ARMSystemLinkage::properties =
 
 void TR::ARMSystemLinkage::initARMRealRegisterLinkage()
    {
-#if 0
-   // Each real register's weight is set to match this linkage convention
-   TR_ARMMachine *machine = cg()->getARMMachine();
-   int icount;
-
-   icount = TR::RealRegister::gr1;
-   machine->getARMRealRegister((TR::RealRegister::RegNum)icount)->setState(TR::RealRegister::Locked);
-   machine->getARMRealRegister((TR::RealRegister::RegNum)icount)->setAssignedRegister(machine->getARMRealRegister((TR::RealRegister::RegNum)icount));
-
-   icount = TR::RealRegister::gr2;
-   machine->getARMRealRegister((TR::RealRegister::RegNum)icount)->setState(TR::RealRegister::Locked);
-   machine->getARMRealRegister((TR::RealRegister::RegNum)icount)->setAssignedRegister(machine->getARMRealRegister((TR::RealRegister::RegNum)icount));
-
-   for (icount=TR::RealRegister::gr3; icount<=TR::RealRegister::gr12; icount++)
-      machine->getARMRealRegister((TR::RealRegister::RegNum)icount)->setWeight(icount);
-
-   for (icount=TR::RealRegister::LastGPR; icount>=TR::RealRegister::gr13; icount--)
-      machine->getARMRealRegister((TR::RealRegister::RegNum)icount)->setWeight(0xf000-icount);
-
-   for (icount=TR::RealRegister::FirstFPR;        icount<=TR::RealRegister::fp3; icount++)
-      machine->getARMRealRegister((TR::RealRegister::RegNum)icount)->setWeight(icount);
+   TR::Machine *machine = cg()->machine();
+
+   // make r15 (PC) unavailable for RA
+   TR::RealRegister *reg = machine->getARMRealRegister(TR::RealRegister::gr15);
+   reg->setState(TR::RealRegister::Locked);
+   reg->setAssignedRegister(reg);
+
+   // make r14 (LR) unavailable for RA
+   reg = machine->getARMRealRegister(TR::RealRegister::gr14);
+   reg->setState(TR::RealRegister::Locked);
+   reg->setAssignedRegister(reg);
+
+   // make r13 (SP) unavailable for RA
+   reg = machine->getARMRealRegister(TR::RealRegister::gr13);
+   reg->setState(TR::RealRegister::Locked);
+   reg->setAssignedRegister(reg);
+
+   // make r12 (IP) unavailable for RA
+   reg = machine->getARMRealRegister(TR::RealRegister::gr12);
+   reg->setState(TR::RealRegister::Locked);
+   reg->setAssignedRegister(reg);
+
+   // make r9 unavailable for RA (just in case, because it's meaning is platform defined)
```

Comment:
Not being an ARM environment expert, is there any way to determine how r9 is intended to be used (e.g., build environment variable, runtime query, etc.) based on the "platform" (in ARM parlance) we're running in?  It would be a shame to default to the most conservative position (preserved and not available for RA) all the time when I suspect it actually is available in most cases.

What does the C runtime expect?